File: password_forge.cpp
// password_forge.cpp
#include <iostream>
#include <string>
#include <vector>
#include <random>
#include <algorithm>
#include <cmath>
#include <fstream>
#include <cctype>
#include <unordered_set>
#include <getopt.h>

using namespace std;

// ANSI colors
const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string YELLOW = "\033[93m";
const string RED = "\033[91m";
const string BLUE = "\033[94m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

// Встроенный словарь (топ-1000 урезанный)
unordered_set<string> common_passwords = {
    "password", "123456", "12345678", "1234", "qwerty", "12345", "dragon",
    "baseball", "football", "letmein", "monkey", "abc123", "mustang", "michael",
    "shadow", "master", "jennifer", "111111", "2000", "jordan", "superman",
    "harley", "1234567", "fuckyou", "hunter", "trustno1", "ranger", "buster",
    "thomas", "tigger", "robert", "soccer", "battle", "killer", "hockey", "george",
    "andrew", "charlie", "donald", "asshole", "qwertyuiop", "liverpool", "nicole",
    "daniel", "ginger", "whatever", "samantha", "jasmine", "passw0rd", "qazwsx"
};

bool is_common_password(const string& pwd) {
    string low = pwd;
    transform(low.begin(), low.end(), low.begin(), ::tolower);
    return common_passwords.find(low) != common_passwords.end();
}

double estimate_entropy(const string& pwd, int charset_size = 0) {
    if (charset_size == 0) {
        unordered_set<char> unique(pwd.begin(), pwd.end());
        charset_size = unique.size();
    }
    if (charset_size <= 1) return 0;
    return pwd.length() * log2(charset_size);
}

string format_time(double seconds) {
    if (seconds == INFINITY) return "∞ (слишком долго)";
    if (seconds < 1) return to_string(seconds) + " секунд";
    else if (seconds < 60) return to_string((int)seconds) + " секунд";
    else if (seconds < 3600) return to_string((int)(seconds/60)) + " минут";
    else if (seconds < 86400) return to_string((int)(seconds/3600)) + " часов";
    else if (seconds < 31536000) return to_string((int)(seconds/86400)) + " дней";
    else return to_string((int)(seconds/31536000)) + " лет";
}

string generate_pronounceable(int length, mt19937& rng) {
    const string vowels = "aeiouy";
    const string consonants = "bcdfghjklmnpqrstvwxz";
    uniform_int_distribution<> vowel_dist(0, vowels.size()-1);
    uniform_int_distribution<> cons_dist(0, consonants.size()-1);
    string pwd;
    for (int i = 0; i < length; ++i) {
        if (i % 2 == 0)
            pwd += consonants[cons_dist(rng)];
        else
            pwd += vowels[vowel_dist(rng)];
    }
    if (length > 0 && (rng() % 2 == 0))
        pwd[0] = toupper(pwd[0]);
    return pwd;
}

string generate_password(int length, bool use_upper, bool use_lower, bool use_digits, bool use_special,
                          bool exclude_ambiguous, bool pronounceable, mt19937& rng) {
    if (pronounceable) return generate_pronounceable(length, rng);
    string chars;
    if (use_lower) chars += "abcdefghijklmnopqrstuvwxyz";
    if (use_upper) chars += "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    if (use_digits) chars += "0123456789";
    if (use_special) chars += "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";
    if (chars.empty()) throw runtime_error("Ни один набор символов не выбран");
    if (exclude_ambiguous) {
        string ambiguous = "O0o1lI";
        for (char c : ambiguous)
            chars.erase(remove(chars.begin(), chars.end(), c), chars.end());
    }
    uniform_int_distribution<> dist(0, chars.size()-1);
    string pwd;
    for (int i = 0; i < length; ++i)
        pwd += chars[dist(rng)];
    return pwd;
}

int main(int argc, char* argv[]) {
    int length = 16, count = 1;
    bool use_upper = true, use_lower = true, use_digits = true, use_special = true;
    bool exclude_ambiguous = false, pronounceable = false, quiet = false;
    double min_entropy = 0;
    string output_file;

    static struct option long_options[] = {
        {"length", required_argument, 0, 'l'},
        {"count", required_argument, 0, 'c'},
        {"uppercase", no_argument, 0, 'u'},
        {"lowercase", no_argument, 0, 'L'},
        {"digits", no_argument, 0, 'd'},
        {"special", no_argument, 0, 's'},
        {"exclude", no_argument, 0, 'e'},
        {"pronounce", no_argument, 0, 'p'},
        {"min-entropy", required_argument, 0, 'm'},
        {"quiet", no_argument, 0, 'q'},
        {"output", required_argument, 0, 'o'},
        {"help", no_argument, 0, 'h'},
        {0,0,0,0}
    };

    int opt;
    while ((opt = getopt_long(argc, argv, "l:c:uLdsep:m:qo:h", long_options, nullptr)) != -1) {
        switch (opt) {
            case 'l': length = stoi(optarg); break;
            case 'c': count = stoi(optarg); break;
            case 'u': use_upper = true; break;
            case 'L': use_lower = true; break;
            case 'd': use_digits = true; break;
            case 's': use_special = true; break;
            case 'e': exclude_ambiguous = true; break;
            case 'p': pronounceable = true; break;
            case 'm': min_entropy = stod(optarg); break;
            case 'q': quiet = true; break;
            case 'o': output_file = optarg; break;
            case 'h':
                cout << "Usage: password_forge [options]\n"
                     << "Options:\n"
                     << "  -l, --length <n>       Длина пароля (default 16)\n"
                     << "  -c, --count <n>        Количество паролей (default 1)\n"
                     << "  -u, --uppercase        Включать заглавные (default on)\n"
                     << "  -L, --lowercase        Включать строчные (default on)\n"
                     << "  -d, --digits           Включать цифры (default on)\n"
                     << "  -s, --special          Включать спецсимволы (default on)\n"
                     << "  -e, --exclude          Исключать неоднозначные символы\n"
                     << "  -p, --pronounce        Генерировать произносимые пароли\n"
                     << "  -m, --min-entropy <n>  Минимальная энтропия (бит)\n"
                     << "  -q, --quiet            Не выводить оценку сложности\n"
                     << "  -o, --output <file>    Сохранить в файл\n"
                     << "  -h, --help             Справка\n";
                return 0;
            default: return 1;
        }
    }

    random_device rd;
    mt19937 rng(rd());

    vector<string> passwords;
    for (int i = 0; i < count; ++i) {
        string pwd;
        int attempts = 0;
        do {
            pwd = generate_password(length, use_upper, use_lower, use_digits, use_special,
                                    exclude_ambiguous, pronounceable, rng);
            attempts++;
            if (attempts > 1000) throw runtime_error("Не удалось сгенерировать пароль с нужной энтропией");
        } while (estimate_entropy(pwd) < min_entropy);
        passwords.push_back(pwd);
    }

    string output;
    for (const auto& pwd : passwords) {
        double entropy = estimate_entropy(pwd);
        bool common = is_common_password(pwd);
        double crack_sec = (entropy == 0) ? INFINITY : pow(2, entropy) / 1e9;
        string crack_str = format_time(crack_sec);
        string strength;
        if (entropy < 30) strength = colorize("Слабый", RED);
        else if (entropy < 50) strength = colorize("Средний", YELLOW);
        else strength = colorize("Сильный", GREEN);

        output += "Пароль: " + pwd + "\n";
        if (!quiet) {
            output += "  Энтропия: " + to_string((int)entropy) + " бит, Стойкость: " + strength;
            output += ", Время взлома: " + crack_str;
            if (common) output += colorize(" (Словарный!)", RED);
            output += "\n\n";
        } else {
            output += "\n";
        }
    }

    if (!output_file.empty()) {
        ofstream f(output_file);
        if (!f) { cerr << colorize("Ошибка записи файла", RED) << endl; return 1; }
        f << output;
        cout << colorize("Результат сохранён в " + output_file, GREEN) << endl;
    } else {
        cout << output;
    }
    return 0;
}

File: password_forge.go
// password_forge.go
package main

import (
	"flag"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strings"
	"time"
)

// ANSI colors
const (
	reset  = "\033[0m"
	green  = "\033[92m"
	yellow = "\033[93m"
	red    = "\033[91m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

// Встроенный словарь
var commonPasswords = map[string]bool{
	"password": true, "123456": true, "12345678": true, "1234": true,
	"qwerty": true, "12345": true, "dragon": true, "baseball": true,
	"football": true, "letmein": true, "monkey": true, "abc123": true,
	"mustang": true, "michael": true, "shadow": true, "master": true,
	"jennifer": true, "111111": true, "2000": true, "jordan": true,
	"superman": true, "harley": true, "1234567": true, "fuckyou": true,
	"hunter": true, "trustno1": true, "ranger": true, "buster": true,
	"thomas": true, "tigger": true, "robert": true, "soccer": true,
	"battle": true, "killer": true, "hockey": true, "george": true,
	"andrew": true, "charlie": true, "donald": true, "asshole": true,
	"qwertyuiop": true, "liverpool": true, "nicole": true, "daniel": true,
	"ginger": true, "whatever": true, "samantha": true, "jasmine": true,
	"passw0rd": true, "qazwsx": true,
}

func isCommon(pwd string) bool {
	return commonPasswords[strings.ToLower(pwd)]
}

func estimateEntropy(pwd string, charsetSize int) float64 {
	if charsetSize == 0 {
		unique := make(map[rune]bool)
		for _, c := range pwd {
			unique[c] = true
		}
		charsetSize = len(unique)
	}
	if charsetSize <= 1 {
		return 0
	}
	return float64(len(pwd)) * math.Log2(float64(charsetSize))
}

func formatTime(seconds float64) string {
	if math.IsInf(seconds, 1) || seconds > 1e30 {
		return "∞ (слишком долго)"
	}
	if seconds < 1 {
		return fmt.Sprintf("%.3f секунд", seconds)
	} else if seconds < 60 {
		return fmt.Sprintf("%.0f секунд", seconds)
	} else if seconds < 3600 {
		return fmt.Sprintf("%.0f минут", seconds/60)
	} else if seconds < 86400 {
		return fmt.Sprintf("%.0f часов", seconds/3600)
	} else if seconds < 31536000 {
		return fmt.Sprintf("%.0f дней", seconds/86400)
	} else {
		return fmt.Sprintf("%.0f лет", seconds/31536000)
	}
}

func generatePronounceable(length int, rng *rand.Rand) string {
	vowels := "aeiouy"
	consonants := "bcdfghjklmnpqrstvwxz"
	pwd := make([]byte, length)
	for i := 0; i < length; i++ {
		if i%2 == 0 {
			pwd[i] = consonants[rng.Intn(len(consonants))]
		} else {
			pwd[i] = vowels[rng.Intn(len(vowels))]
		}
	}
	if rng.Intn(2) == 0 && length > 0 {
		pwd[0] = byte(strings.ToUpper(string(pwd[0]))[0])
	}
	return string(pwd)
}

func generatePassword(length int, useUpper, useLower, useDigits, useSpecial bool,
	excludeAmbiguous, pronounceable bool, rng *rand.Rand) string {
	if pronounceable {
		return generatePronounceable(length, rng)
	}
	var chars []rune
	if useLower {
		chars = append(chars, []rune("abcdefghijklmnopqrstuvwxyz")...)
	}
	if useUpper {
		chars = append(chars, []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ")...)
	}
	if useDigits {
		chars = append(chars, []rune("0123456789")...)
	}
	if useSpecial {
		chars = append(chars, []rune("!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~")...)
	}
	if len(chars) == 0 {
		panic("Ни один набор символов не выбран")
	}
	if excludeAmbiguous {
		ambiguous := map[rune]bool{'O': true, '0': true, 'o': true, '1': true, 'l': true, 'I': true}
		filtered := make([]rune, 0)
		for _, c := range chars {
			if !ambiguous[c] {
				filtered = append(filtered, c)
			}
		}
		chars = filtered
	}
	pwd := make([]rune, length)
	for i := 0; i < length; i++ {
		pwd[i] = chars[rng.Intn(len(chars))]
	}
	return string(pwd)
}

func main() {
	var length int
	var count int
	var uppercase bool
	var lowercase bool
	var digits bool
	var special bool
	var exclude bool
	var pronounce bool
	var minEntropy float64
	var quiet bool
	var output string
	var help bool

	flag.IntVar(&length, "l", 16, "Длина пароля")
	flag.IntVar(&count, "c", 1, "Количество паролей")
	flag.BoolVar(&uppercase, "u", true, "Включать заглавные")
	flag.BoolVar(&lowercase, "L", true, "Включать строчные")
	flag.BoolVar(&digits, "d", true, "Включать цифры")
	flag.BoolVar(&special, "s", true, "Включать спецсимволы")
	flag.BoolVar(&exclude, "e", false, "Исключать неоднозначные")
	flag.BoolVar(&pronounce, "p", false, "Произносимый режим")
	flag.Float64Var(&minEntropy, "m", 0, "Минимальная энтропия (бит)")
	flag.BoolVar(&quiet, "q", false, "Не выводить оценку")
	flag.StringVar(&output, "o", "", "Файл для сохранения")
	flag.BoolVar(&help, "h", false, "Справка")
	flag.Parse()

	if help {
		fmt.Println(`Usage: password_forge [options]
Options:
  -l <n>       Длина пароля (default 16)
  -c <n>       Количество паролей (default 1)
  -u           Включать заглавные (default on)
  -L           Включать строчные (default on)
  -d           Включать цифры (default on)
  -s           Включать спецсимволы (default on)
  -e           Исключать неоднозначные символы
  -p           Генерировать произносимые пароли
  -m <n>       Минимальная энтропия (бит)
  -q           Не выводить оценку сложности
  -o <file>    Сохранить в файл
  -h           Справка`)
		return
	}

	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	var passwords []string
	for i := 0; i < count; i++ {
		var pwd string
		attempts := 0
		for {
			pwd = generatePassword(length, uppercase, lowercase, digits, special,
				exclude, pronounce, rng)
			attempts++
			if attempts > 1000 {
				panic("Не удалось сгенерировать пароль с нужной энтропией")
			}
			if estimateEntropy(pwd, 0) >= minEntropy {
				break
			}
		}
		passwords = append(passwords, pwd)
	}

	var outputText string
	for _, pwd := range passwords {
		entropy := estimateEntropy(pwd, 0)
		common := isCommon(pwd)
		crackSec := math.Pow(2, entropy) / 1e9
		if entropy == 0 {
			crackSec = math.Inf(1)
		}
		crackStr := formatTime(crackSec)
		var strength string
		if entropy < 30 {
			strength = colorize("Слабый", red)
		} else if entropy < 50 {
			strength = colorize("Средний", yellow)
		} else {
			strength = colorize("Сильный", green)
		}

		outputText += fmt.Sprintf("Пароль: %s\n", pwd)
		if !quiet {
			outputText += fmt.Sprintf("  Энтропия: %.0f бит, Стойкость: %s", entropy, strength)
			outputText += fmt.Sprintf(", Время взлома: %s", crackStr)
			if common {
				outputText += colorize(" (Словарный!)", red)
			}
			outputText += "\n\n"
		} else {
			outputText += "\n"
		}
	}

	if output != "" {
		err := os.WriteFile(output, []byte(outputText), 0644)
		if err != nil {
			fmt.Println(colorize("Ошибка записи файла: "+err.Error(), red))
			os.Exit(1)
		}
		fmt.Println(colorize("Результат сохранён в "+output, green))
	} else {
		fmt.Print(outputText)
	}
}
